keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+> * `additionalItems`,
+> * `properties`,
+> * `patternProperties`,
+> * `additionalProperties`,
+> * `default`,
+> giving `drop(s)`,
+> 2. then merging `s_i` into it’s containing object.
+>
+>    I.e.: `skel(s) := merge(drop(s), s_1, …, s_n)` with
+>
+>```
+>merge(x_1, …, x_n) := { 
+>   "type": t if all x_i agree on t as type, undefined otherwise
+>   "items": [ merge(x_i1, …, x_ik) ] where s_ij = s_j.items[i] if defined,
+>   "additionaItems": merge(p_1, …, p_i),
+>       for all x_i with defined additionalItems p_i
+>   "properties": { k_i: merge(v_i1, …, v_ik) for keys appearing in x_ij with values x_ij },
+>   "patternProperties": { k_i: merge(v_i1, …, v_ik) for keys appearing in x_ij with values x_ij },
+>   "additionalProperties": merge(p_1, …, p_i),
+>       for all x_i with defined additionalProperties p_i
+>   "default": first default value of  `x_i` or fail
+>}
+>```
+
+The skeleton schema especially lacks:
+* all value validations like `pattern`, `format`, `minValue`, `maxValue`, ...
+* all propositional operators like `anyOf`, `allOf`, `oneOf`, `not`.
+
+The skeleton schema `skel(s)` puts
+* less or equal constraints on `type` than `s`.
+* every field specified by a `properties` key in `s` is specified in `skel(s)`.
+* if there are no default values under `allOf`, `anyOf`, `oneOf`, `not` in `s`, for each JSON path the default matches in `s` and `skel(s)`.
+
+The computation of `skel(s)` is `O(size(s))` and `size(skel(s)) <= size(s)`.
+
+Note that a field might be constrained by the `type` construct in the full OpenAPI validation schema, but not in its skeleton. This is fine because we have to support polymorphic fields like `IntOrString`, but avoid  `allOf`, `anyOf`, `oneOf`, `not` in the skeleton.
+
+**Property:** if the OpenAPI validation schema applies to an object, so does its skeleton schema.
+
+Pruning and defaulting will be implemented based on the skeleton schema of the specified OpenAPI validation schema.
+
+Optionally for debugging, the skeleton schema derived from the validation schema by the apiextensions-apiserver can be stored in the CRD status.
+
+### Example 8
+
+>Assume 
+>
+>```
+>s := {
+>    "anyOf": [
+>        {"properties": {"a": {"type": "string"}}},
+>        {"properties": {"a": {"type": "integer"}, "b": {"type": "string"}}}
+>    ],
+>    "properties": {"c": {}}
+>}
+>```
+>
+>Then 
+>
+>```
+>skel(s) = {
+>    "properties": {
+>	    "a": {},
+>	    "b": {"type": "string"},
+>	    "c": {}
+>    }
+>}
+>```
+>
+> The `type` of `a` does not match on all paths. Hence, it is omitted in the skeleton. In contrast, `b` has a unique `type` of `"string"`, so it stays in the skeleton.
+
+### Types and Formats
+
+Note, that having less `type` constraints in the skeleton than in the whole OpenAPI validation schema means that admission and conversion will see possibly wrong types. The final validation in the registry validation phase will check for the complete OpenAPI validation schema and catch those type errors.
```

Comment:
**Mehdy Bohlool:** so the plan is to pass `skel` schema to these webhooks instead of the full schema?

**Stefan Schimanski:** no, the webhooks won't see the schema at all. It is embedded in the decoder. But the webhooks will see the result of decoding, and without type+format the result might be unexpected in the sense that the Golang JSON decoder used by the webhook implementation – if that uses a Golang representation of CustomResources – will fail.